File: skyspark-scala/src/main/scala/com/akikanellis/skyspark/scala/api/algorithms/bnl/FlagAdder.scala
package com.akikanellis.skyspark.scala.api.algorithms.bnl

import com.akikanellis.skyspark.scala.api.algorithms.Point
import org.apache.spark.rdd.RDD

/**
  * The Map task of the Division part of the MR-BNL algorithm.
  * <p>
  * Adds the corresponding flags to the points.
  *
  * @param medianFinder The median of the points
  */
private[bnl] class FlagAdder(private val medianFinder: MedianFinder) extends Serializable {

  private[bnl] def this() = this(new MedianFinder)

  private[bnl] def addFlags(points: RDD[Point]): RDD[(Flag, Point)] = {
    val median = medianFinder.getMedian(points)
    val flagProducer = createFlagProducer(median)

    points.map(p => (flagProducer.calculateFlag(p), p))
  }

  private[bnl] def createFlagProducer(median: Point): FlagProducer = new FlagProducer(median)
}

File: skyspark-java/src/main/java/com/akikanellis/skyspark/api/algorithms/bnl/FlagAdder.java
package com.akikanellis.skyspark.api.algorithms.bnl;

import com.akikanellis.skyspark.api.algorithms.Point;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import scala.Tuple2;

import java.io.Serializable;

class FlagAdder implements Serializable {
    private final MedianFinder medianFinder;

    FlagAdder(MedianFinder medianFinder) { this.medianFinder = medianFinder; }

    JavaPairRDD<Flag, Point> addFlags(JavaRDD<Point> points) {
        Point median = medianFinder.getMedian(points);
        FlagProducer flagProducer = createFlagProducer(median);

        return points.mapToPair(p -> new Tuple2<>(flagProducer.calculateFlag(p), p));
    }

    protected FlagProducer createFlagProducer(Point median) { return new FlagProducer(median); }
}
